File: event-service/src/main/java/com/github/manosbatsis/services/event/model/UserEventKey.java
package com.github.manosbatsis.services.event.model;

import java.util.Date;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.cassandra.core.cql.Ordering;
import org.springframework.data.cassandra.core.cql.PrimaryKeyType;
import org.springframework.data.cassandra.core.mapping.PrimaryKeyClass;
import org.springframework.data.cassandra.core.mapping.PrimaryKeyColumn;

@Data
@AllArgsConstructor
@NoArgsConstructor
@PrimaryKeyClass
public class UserEventKey {

  @PrimaryKeyColumn(ordinal = 0, type = PrimaryKeyType.PARTITIONED)
  private Long userId;

  @PrimaryKeyColumn(ordinal = 1, type = PrimaryKeyType.CLUSTERED, ordering = Ordering.ASCENDING)
  private Date datetime;
}

File: event-service/src/main/kotlin/com/github/manosbatsis/services/event/model/UserEventKey.kt
package com.github.manosbatsis.services.event.model

import org.springframework.data.cassandra.core.cql.Ordering
import org.springframework.data.cassandra.core.cql.PrimaryKeyType
import org.springframework.data.cassandra.core.mapping.PrimaryKeyClass
import org.springframework.data.cassandra.core.mapping.PrimaryKeyColumn
import java.util.Date

@PrimaryKeyClass
data class UserEventKey(
    @field:PrimaryKeyColumn(ordinal = 0, type = PrimaryKeyType.PARTITIONED)
    val userId: Long,
    @field:PrimaryKeyColumn(ordinal = 1, type = PrimaryKeyType.CLUSTERED, ordering = Ordering.ASCENDING)
    val datetime: Date,
) {
    fun asString() = userId + datetime.time
}
